### PR TITLE
imxrt-multi/uart: add intermediate rx fifo between irq handler and thread

### PR DESCRIPTION
**Pull request by gerard5**
#### imxrt-multi/uart: add intermediate rxfifo (irq-thread)
    
The rx overrun flag is cleared in the interrupt handler. The hardware fifo turns out to be too small (four octets) and cannot be increased (read only register on imxrt1176). RX bytes were lost in irq handler. To solve the issue this change adds small FIFO between the interrupt and the thread to improve performance significantly. "Lock-free FIFO" from libtty implementation used to accomplish this.
    
Additionally, statistics regarding the counting of (software and hardware) overrun events are introduced.

Removed unused variable in uart_common, and corrected clock settings for imxrt1176 to behave like on imxrt10xx (just make clock run).

#### imxrt-multi/uart: allow per instance UARTx_BUFSIZE

This change allows to set libtty buffers for each uart instance separately. By default, the UART_BUFSIZE setting for an instance will be overridden by the UARTx_BUFSIZE (x - instance number) setting when defined in the board configuration.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: `imxrt1176-nil`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: multi/imxrt-multi/uart.c [hunk 1]
```diff
@@ -86,52 +97,66 @@ static const int uartPos[] = { UART1_POS, UART2_POS, UART3_POS, UART4_POS, UART5
...
+		}
+	}
 
+	/* Clear errors: parity, framing, noise, overrun and idle flag */
+	*(uart->base + statr) = (status & (0x1fuL << 16));
 
-static inline int uart_getTXcount(uart_t *uart)
-{
-	return (*(uart->base + waterr) >> 8) & 0xff;
+	/* reschedule */
```
Comment (agkaminski) [inline multi/imxrt-multi/uart.c:138]: Nitpick - not exactly, wakeup registered cond (uart->cond)

File: multi/imxrt-multi/uart.c [hunk 2]
```diff
@@ -67,12 +69,21 @@ typedef struct uart_s {
...
+
+	/* fifo between irq and thread */
+	lf_fifo_t rxFifoCtx;
+	uint8_t rxFifoData[UART_RXFIFOSIZE];
+
+	/* statistics */
+	struct {
+		size_t hw_overrunCntr;
+		size_t sw_overrunCntr;
+	} stat;
```
Comment (agkaminski) [inline multi/imxrt-multi/uart.c:81]: This struct seems write-only

Comment (gerard5) [reply multi/imxrt-multi/uart.c:81]: imxrt-multi does not have mechanisms for reading counters (yet), this structure is helpful for reading with a debugger. I can add a comment.

File: multi/imxrt-multi/uart.c (hunk 1 above)
Comment (gerard5) [reply multi/imxrt-multi/uart.c]: Right, thanks.

**Review by agkaminski — APPROVED** [review]

**Pull request merged**